File: ScalaQuickStart/src/main/java/com/tirthal/learning/java2scala/javaway/inheritance/SubTypeSample.java
package com.tirthal.learning.java2scala.javaway.inheritance;

/**
 * Test class for sub-type inheritance demo - Dog class is inheriting behavior of Animal class and can have its own too.
 *
 * @author tirthalp
 * @see <a href="https://github.com/tirthalpatel/Learning-OOPD/blob/master/DesignPrinciples/src/com/tirthal/learning/oop/concepts/Inheritance_TestDrive.java">Inheritance Concept</a>
 */
public class SubTypeSample {

	public static void main(String[] args) {
		Dog dog = new Dog(123);
		dog.sound();
	}
}

// Super class
class Animal {
	private final int id;

	public Animal(int id) {
		this.id = id;
	}

	public int getId() {
		return this.id;
	}
	public void sound() {
		System.out.println("By default it is mute");
	}
}

// Sub class
class Dog extends Animal {

	public Dog(int id) {
		super(id);
	}

	// Own behavior
	private void bark() {
		System.out.println("Dog '" + getId() + "' is barking");
	}

	// Overriding super class behavior
	@Override
	public void sound() {
		bark();
	}
}

File: ScalaQuickStart/src/main/scala/com/tirthal/learning/java2scala/scalaway/inheritance/SubTypeSample.scala
package com.tirthal.learning.java2scala.scalaway.inheritance

//  Test class instance for sub-type inheritance demo - Dog class is inheriting behavior of Animal class and can have its own too.
object SubTypeSample {
  def main(args: Array[String]) {
	val d1 = new Dog(123)
	d1.sound()
  }
}

// Super class
class Animal(val id: Int) {
  // val keyword in primary constructor of Animal class = instructs scala compiler to create fields and public getter access methods (not creating setter methods).

  // generic method of super class
  def sound() = println("By default it is mute")
}

// Sub class
class Dog(id: Int) extends Animal(id) {

  // Own behavior
  private def bark() = println("Dog '" + id + "' is barking")

  // Overriding super class behavior
  // In Java - @Override is optional
  // In Scala - It is mandatory to declare method as "override", when you override concrete (non-abstract) method
  override def sound() = bark()
}